### PR TITLE
[FIXED] RACE in Sublist

### LINKED ISSUE #640: DATA RACE when going over results of sublist match and sub removed from sublist
This was seen when working on PR #638, and showed on an unrelated test. The new test added in the PR (TestRoutedQueueUnsubscribe) is reporting race at almost every run when running locally.
```
IvanMBP:gnatsd ivan$ go test -race -v -run=TestRoutedQueueUnsubscribe -count 10 ./server/
=== RUN   TestRoutedQueueUnsubscribe
==================
WARNING: DATA RACE
Write at 0x00c420496af8 by goroutine 31:
  github.com/nats-io/gnatsd/server.removeSubFromList()
      /Users/ivan/dev/go/src/github.com/nats-io/gnatsd/server/sublist.go:440 +0x100
  github.com/nats-io/gnatsd/server.(*Sublist).removeFromNode()
      /Users/ivan/dev/go/src/github.com/nats-io/gnatsd/server/sublist.go:460 +0x174
  github.com/nats-io/gnatsd/server.(*Sublist).Remove()
      /Users/ivan/dev/go/src/github.com/nats-io/gnatsd/server/sublist.go:378 +0x80a
  github.com/nats-io/gnatsd/server.(*client).unsubscribe()
      /Users/ivan/dev/go/src/github.com/nats-io/gnatsd/server/client.go:847 +0x27d
  github.com/nats-io/gnatsd/server.(*client).deliverMsg()
      /Users/ivan/dev/go/src/github.com/nats-io/gnatsd/server/client.go:1002 +0x708
  github.com/nats-io/gnatsd/server.(*client).processMsg()
      /Users/ivan/dev/go/src/github.com/nats-io/gnatsd/server/client.go:1142 +0x20b0
  github.com/nats-io/gnatsd/server.(*client).parse()
      /Users/ivan/dev/go/src/github.com/nats-io/gnatsd/server/parser.go:226 +0x1fd1
  github.com/nats-io/gnatsd/server.(*client).readLoop()
      /Users/ivan/dev/go/src/github.com/nats-io/gnatsd/server/client.go:284 +0xe93
  github.com/nats-io/gnatsd/server.(*Server).createRoute.func2()
      /Users/ivan/dev/go/src/github.com/nats-io/gnatsd/server/route.go:407 +0x41

Previous read at 0x00c420496af8 by goroutine 43:
  github.com/nats-io/gnatsd/server.(*client).processMsg()
      /Users/ivan/dev/go/src/github.com/nats-io/gnatsd/server/client.go:1198 +0x1296
  github.com/nats-io/gnatsd/server.(*client).parse()
      /Users/ivan/dev/go/src/github.com/nats-io/gnatsd/server/parser.go:226 +0x1fd1
  github.com/nats-io/gnatsd/server.(*client).readLoop()
      /Users/ivan/dev/go/src/github.com/nats-io/gnatsd/server/client.go:284 +0xe93
  github.com/nats-io/gnatsd/server.(*Server).createClient.func2()
      /Users/ivan/dev/go/src/github.com/nats-io/gnatsd/server/server.go:805 +0x41

Goroutine 31 (running) created at:
  github.com/nats-io/gnatsd/server.(*Server).startGoRoutine()
      /Users/ivan/dev/go/src/github.com/nats-io/gnatsd/server/server.go:1059 +0xba
  github.com/nats-io/gnatsd/server.(*Server).createRoute()
      /Users/ivan/dev/go/src/github.com/nats-io/gnatsd/server/route.go:407 +0x62a
  github.com/nats-io/gnatsd/server.(*Server).connectToRoute()
      /Users/ivan/dev/go/src/github.com/nats-io/gnatsd/server/route.go:784 +0x4cc
  github.com/nats-io/gnatsd/server.(*Server).solicitRoutes.func1()
      /Users/ivan/dev/go/src/github.com/nats-io/gnatsd/server/route.go:798 +0x4f

Goroutine 43 (running) created at:
  github.com/nats-io/gnatsd/server.(*Server).startGoRoutine()
      /Users/ivan/dev/go/src/github.com/nats-io/gnatsd/server/server.go:1059 +0xba
  github.com/nats-io/gnatsd/server.(*Server).createClient()
      /Users/ivan/dev/go/src/github.com/nats-io/gnatsd/server/server.go:805 +0x673
  github.com/nats-io/gnatsd/server.(*Server).AcceptLoop.func2()
      /Users/ivan/dev/go/src/github.com/nats-io/gnatsd/server/server.go:475 +0x58
==================
If this test is brought to master, the same issue happens, so this is not specific to that branch.

### PR DESCRIPTION
**Pull request by kozlovic**
This would manifest for instance when server tries to send messages
to queue subscribers and a subscription is unsubsribed at the same
time.

Resolves #640

### CODE REVIEW

**Comment by kozlovic** [comment]
@derekcollison Not sure if this the only way to fix, but clearly there was an issue between the results from a match and a subscription being removed from the sublist.

File: server/sublist.go [hunk 1]
```diff
@@ -236,8 +236,13 @@ func (s *Sublist) Match(subject string) *SublistResult {
 	s.Lock()
 	matchLevel(s.root, tokens, result)
 
+	// Store in cache and return to caller a copy of the results to avoid
+	// race when sub is removed from sublist and caller walks through the
+	// results.
+	cr := copyResult(result)
```
Comment (derekcollison) [inline server/sublist.go]: This feels like it could be a performance hit for sure.

File: server/sublist_test.go [hunk 2]
```diff
@@ -1,3 +1,6 @@
+// Copyright 2012-2017 Apcera Inc. All rights reserved.
+// Copyright 2018 Synadia Communications Inc. All rights reserved.
```
Comment (derekcollison) [inline server/sublist_test.go]: I will change all of these at some point next week.

File: server/sublist_test.go [hunk 3]
```diff
@@ -509,6 +512,81 @@ func TestSublistRemoveWithWildcardsAsLiterals(t *testing.T) {
...
+		go func() {
+			for _, sub := range subs {
+				s.Remove(sub)
+			}
+			wg.Done()
+		}()
+		for _, qsub := range r.qsubs {
+			for i := 0; i < len(qsub); i++ {
+				sub := qsub[i]
+				if string(sub.queue) != "bar" {
```
Comment (derekcollison) [inline server/sublist_test.go]: Should we rewrite it above to make it deterministic?

File: server/sublist.go (hunk 1 above)
Comment (kozlovic) [reply server/sublist.go]: I revised to just copy where needed (in matchLevel)

File: server/sublist_test.go (hunk 3 above)
Comment (kozlovic) [reply server/sublist_test.go]: Not sure what you mean here?

File: server/sublist.go [hunk 4]
```diff
@@ -266,7 +261,9 @@ func addNodeToResults(n *node, results *SublistResult) {
 		if i := findQSliceForSub(qr[0], results.qsubs); i >= 0 {
 			results.qsubs[i] = append(results.qsubs[i], qr...)
 		} else {
-			results.qsubs = append(results.qsubs, qr)
+			copyqr := make([]*subscription, len(qr))
+			copy(copyqr, qr)
+			results.qsubs = append(results.qsubs, copyqr)
```
Comment (derekcollison) [inline server/sublist.go]: Any noticeable performance hit? Any other way to avoid the allocation and copy even of the array?

**Pull request merged**